File: lab07/lab07.эйлер-коши.py
import numpy as np
import math

def euler_method(f, y0, x0, x_end, h):
    n = int((x_end - x0) / h)
    x = np.linspace(x0, x_end, n+1)
    y = np.zeros(n+1)
    y[0] = y0
    for i in range(n):
        y[i+1] = y[i] + h * f(x[i], y[i])
    return x, y

def f1(x, y):
    return -y * math.tan(x) + math.sin(2*x)

x0, y0 = 0, -1
x_end, h = 1, 0.1

x, y = euler_method(f1, y0, x0, x_end, h)

for xi, yi in zip(x, y):
    print(f"x: {xi:.1f}, y: {yi:.4f}")

File: lab07/lab07.эйлер-коши.cpp
#include <iostream>
#include <vector>
#include <cmath>

double f1(double x, double y) {
    return -y * tan(x) + sin(2 * x);
}

std::vector<double> euler_method(double (*f)(double, double), double y0, double x0, double x_end, double h) {
    int n = (x_end - x0) / h;
    std::vector<double> x(n+1), y(n+1);
    x[0] = x0;
    y[0] = y0;
    for (int i = 0; i < n; ++i) {
        x[i+1] = x[i] + h;
        y[i+1] = y[i] + h * f(x[i], y[i]);
    }
    return y;
}

int main() {
    double x0 = 0, y0 = -1, x_end = 1, h = 0.1;
    auto y = euler_method(f1, y0, x0, x_end, h);

    for (int i = 0; i < y.size(); ++i) {
        std::cout << "x: " << x0 + i*h << ", y: " << y[i] << std::endl;
    }

    return 0;
}
